compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+                            slots.push(s);
+                        }
+                    }
+                    StatementKind::StorageDead(l) => {
+                        if locals.contains(&l) {
+                            storage_dead.push(statement.replace_nop());
+                            slots.push(s);
+                        }
+                    }
+                    _ => {}
+                }
+            }
+            slots.sort();
+
+            assert!(slots.len() >= storage_live.len() + 1 + storage_dead.len());
+            for (slot, statement) in slots.iter().zip(storage_live.into_iter().chain([statement])) {
+                assert!(matches!(statements[*slot].kind, StatementKind::Nop));
+                statements[*slot] = statement;
+            }
+            for (slot, statement) in slots.iter().rev().zip(storage_dead.into_iter().rev()) {
+                assert!(matches!(statements[*slot].kind, StatementKind::Nop));
+                statements[*slot] = statement;
+            }
+
+            continue 'outer;
+        }
+    }
+
+    !invalidated_statements.is_empty()
+}
+
+fn find_locals(statement: &Statement<'_>) -> SmallVec<[Local; 4]> {
+    struct LocalCollector {
+        locals: SmallVec<[Local; 4]>,
+    }
+    impl Visitor<'_> for LocalCollector {
+        fn visit_local(&mut self, local: Local, _context: PlaceContext, _location: Location) {
+            self.locals.push(local);
+        }
+    }
+
+    let mut visitor = LocalCollector { locals: SmallVec::new() };
+    visitor.visit_statement(statement, Location::START);
+    visitor.locals
+}
+
+fn as_assign<'a, 'tcx>(
+    result: Set1<Location>,
+    body: &'a Body<'tcx>,
+) -> Option<(Location, &'a (Place<'tcx>, Rvalue<'tcx>))> {
+    let location = match result {
+        Set1::One(location) => location,
+        Set1::Empty | Set1::Many => return None,
+    };
+
+    body.stmt_at(location).left()?.kind.as_assign().map(|res| (location, res))
+}
+
+#[derive(Debug, Clone)]
+struct Analysis {
+    read: Set1<Location>,
+    write: Set1<Location>,
+}
+
+impl Default for Analysis {
+    fn default() -> Self {
+        Self { read: Set1::Empty, write: Set1::Empty }
+    }
+}
+
+struct AnalysisVisitor {
+    analysis: IndexVec<Local, Analysis>,
+}
+
+impl<'tcx> Visitor<'tcx> for AnalysisVisitor {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        use rustc_middle::mir::visit::MutatingUseContext;
+        match context {
+            PlaceContext::NonUse(_) => {}
+            PlaceContext::MutatingUse(MutatingUseContext::Store) => {
+                self.analysis[place.local].write.insert(location);
+            }
+            PlaceContext::MutatingUse(
+                MutatingUseContext::Borrow
+                | MutatingUseContext::Projection
+                | MutatingUseContext::AddressOf,
+            ) => {
+                self.analysis[place.local].read.insert(location);
+            }
```

Comment:
This branch needs an explanation.
We usually consider all refs and addressofs, either mutable or immutable, to be writes in optimizations.
Why are they all reads here?